tests/lib/bin/session-tool

```diff
@@ -0,0 +1,162 @@
+			;;
+		*)
+			break
+			;;
+	esac
+done
+
+# This fixes a bug in some older Debian systems where /root/.profile contains
+# unconditional invocation of mesg, which on non-tty shells prints "mesg
+# ttyname failed inappropriate ioctl for device" which pollutes output from
+# invoked programs.
+# TODO: move this to spread wide project setup.
+sed -i -e 's/mesg n .*true/tty -s \&\& mesg n/g' /root/.profile
+
+read -r uuid < /proc/sys/kernel/random/uuid
+unit_name="session-tool-$uuid.service"
+tmp_dir="/tmp/session-tool-$uuid"
+mkdir "$tmp_dir"
+mkfifo -m 0666 "$tmp_dir/dbus-monitor.pipe" "$tmp_dir/ready.pipe" "$tmp_dir/result.pipe" "$tmp_dir/stdin.pipe" "$tmp_dir/stdout.pipe" "$tmp_dir/stderr.pipe"
+trap 'rm -rf $tmp_dir' EXIT
+
+# Run dbus-monitor waiting for systemd JobRemoved signal. Process the output
+# with awk, forwarding the result of the service to a fifo.
+monitor_expr="type='signal', sender='org.freedesktop.systemd1', interface='org.freedesktop.systemd1.Manager', path='/org/freedesktop/systemd1', member='JobRemoved'"
+stdbuf -oL dbus-monitor --system --monitor "$monitor_expr" > "$tmp_dir/dbus-monitor.pipe" &
+dbus_monitor_pid=$!
+
+awk_expr="
+BEGIN {
+	found=0;
+	ready=0;
+}
+
+# Once we get the NameAcquired message we are sure dbus-monitor is connected
+# and will receive JobRemoved once it is sent. The reason we are getting this
+# message despite the filter we establish above is that it is sent directly to
+# us, which bypasses DBus filter expressions.
+/member=NameAcquired/ {
+	if (!ready) {
+		ready=1;
+		print ready > \"$tmp_dir/ready.pipe\";
+		close(\"$tmp_dir/ready.pipe\");
+	}
+}
+
+# This part matches an argument to JobRemoved that contains the name of the
+# session-tool-xxx.service name we picked earlier. Once we see this we are sure
+# the job is gone.
+/   string \"$unit_name\"/ {
+	found=1;
+	print \"found service file\";
+	fflush();
+	next;
+}
+
+# This matches any string argument but only takes effect once we found our
+# JobRemoved message. The order of arguments to JobRemoved is such that the
+# immediate successor of the service name is the result word. This is
+# translated to a pass / fail exit code from session-tool. Scanning this part
+# terminates awk.
+/string \".*\"/ {
+	if (found==1) {
+		print \"found result\";
+		print \$2 > \"$tmp_dir/result.pipe\";
+		fflush(\"\");
+		close(\"$tmp_dir/result.pipe\");
+		exit;
+	}
+}
+"
+awk -W interactive "$awk_expr" <"$tmp_dir/dbus-monitor.pipe" >"$tmp_dir/awk.log" 2>&1 &
+awk_pid=$!
+
+# Wait for dbus-monitor to start.
+cat "$tmp_dir/ready.pipe" >/dev/null
+
+# Use busctl to spawn a command. The command is wrapped in shell, runuser -l
+# and redirects to capture output. Sadly busctl doesn't support passing file
+# descriptors https://github.com/systemd/systemd/issues/14954 As a workaround
+# we pass a set of pipes. This is good for non-interactive work.
+cat >"$tmp_dir/stdin.pipe" &
+cat <"$tmp_dir/stdout.pipe" >&1 &
+cat <"$tmp_dir/stderr.pipe" >&2 &
+busctl \
```

Comment:
How close is this to what systemd-run does?

Reply:
It's relatively close. It's pretty much what `systems-run --wait --pipe` does but we cannot rely on that (because it's only available in recent systemd version).